src/main/kotlin/controller/IssueController.kt

```diff
@@ -19,52 +22,126 @@ import tornadofx.asObservable
+        filter.value = newFilter
+        applyFilter(newFilter)
     }
 
-    @Suppress("IMPLICIT_CAST_TO_ANY")
-    suspend fun recordTime(issueWithTime: IssueWithTime) {
-        val credentials = credentialController.credentials ?: throw NoCredentialsError()
+    /**
+     * Records time for an issue.
+     *
+     * @return A result stating whether or not the recording was successful, and if not why
+     */
+    suspend fun recordTime(issueWithTime: IssueWithTime): TimeRecordResult {
+        val credentials = credentialController.credentials ?: return TimeRecordResult.NoCredentials
         val success = GitlabAPI.issue.addTimeSpentToIssue(credentials, issueWithTime.issue.projectID, issueWithTime.issue.idInProject, issueWithTime.elapsedTime.toString())
 
-        if (success) {
+        return if (success) {
             val updatedIssue = issueWithTime.issue.copy(timeSpent = issueWithTime.issue.timeSpent + issueWithTime.elapsedTime)
 
             withContext(Dispatchers.JavaFx) {
                 val issueIdx = issueList.indexOf(issueWithTime.issue)
 
-                if (issueIdx == -1) {
-                    issueList.add(updatedIssue)
-                } else {
-                    issueList.set(issueIdx, updatedIssue)
+                if (issueIdx != -1) {
+                    issueList[issueIdx] = updatedIssue
                 }
             }
+
+            TimeRecordResult.TimeRecorded
+        } else {
+            TimeRecordResult.TimeFailedToRecord
+        }
+    }
+
+    /**
+     * Using data already available on the unfiltered list, locally filters the unfiltered issue list
+     */
+    private suspend fun applyFilter(filter: IssueFilter) {
+        // Pulling this in in the event something else modifies the list
+        val currentIssueList = unfilteredIssueList
+
+        // First filter by milestone
+        val milestoneFilteredIssues = when (filter.selectedMilestone) {
+            is MilestoneFilterOption.NoMilestoneOptionSelected -> currentIssueList.asSequence()
+            is MilestoneFilterOption.HasAssignedMilestone -> currentIssueList.asSequence().filter { it.milestone != null }
+            is MilestoneFilterOption.HasNoMilestone -> currentIssueList.asSequence().filter { it.milestone == null }
+            is MilestoneFilterOption.SelectedMilestone -> currentIssueList.asSequence().filter { it.milestone?.idInProject == filter.selectedMilestone.milestone.idInProject }
+        }
+
+        // Next filter by filter text
+        val filteredIssues = if (filter.filterText.isEmpty()) {
+            milestoneFilteredIssues
+        } else {
+            milestoneFilteredIssues.filter { issue ->
+                val searchText = "#${issue.idInProject} ${issue.title}"
+                searchText.contains(filter.filterText, ignoreCase = true)
+            }
+        }
+
+        withContext(Dispatchers.JavaFx) {
+            issueList.setAll(filteredIssues.toList())
```

Comment:
The rest of this function should be on `Dispatchers.Default`, including the `toList()` call here. Once the value is computed we can set the issue list on the JavaFX dispatcher